pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,218 @@
+		return nil, err
+	}
+
+	tokenGenericObj, err := r.accessTokenStorage.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	token := tokenGenericObj.(*oauthapi.OAuthAccessToken)
+	if token == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if !isValidUserToken(token, ctxUserName) {
+		return nil, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	return (*oauthapi.UserOAuthAccessToken)(token), nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher, ctxUserName)
+	go watcher.Watch(ctx)
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	var deletedRV string
+	if options.Preconditions != nil && options.Preconditions.ResourceVersion != nil {
+		deletedRV = *options.Preconditions.ResourceVersion
+	}
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return nil, false, err
+	}
+
+	deletedTyped := deletedObj.(*oauthapi.UserOAuthAccessToken)
+	if !isValidUserToken((*oauthapi.OAuthAccessToken)(deletedTyped), ctxUserName) {
+		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	// don't mutate the original options
+	newOpts := options.DeepCopy()
+	if newOpts.Preconditions == nil {
+		newOpts.Preconditions = &metav1.Preconditions{UID: &deletedTyped.UID}
+	} else if newOpts.Preconditions.UID == nil {
+		newOpts.Preconditions.UID = &deletedTyped.UID
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, newOpts)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		klog.V(4).Infof("no user found in the context")
+		return "", kerrors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
+	}
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		klog.V(4).Info("user in context has zero-length username")
```

Comment:
same here. No username means nothing to see. `return "", false, nil` and NotFound where this func is called.

Reply:
what's the use of error, then?

Reply:
right, just get rid of it. I prefer to put logic for errors returned to the user at the top-level, not deep into helpers.